<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>High-throughput phenotyping applied to plant breeding</title>
    <meta charset="utf-8" />
    <meta name="author" content="Rafael Massahiro Yassue" />
    <meta name="date" content="2025-09-03" />
    <script src="libs/header-attrs-2.29/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/default-fonts.css" rel="stylesheet" />
    <script src="libs/kePrint-0.0.1/kePrint.js"></script>
    <link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
    <link rel="stylesheet" href="animate.css" type="text/css" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/animate.css/4.1.1/animate.min.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

.title[
# High-throughput phenotyping applied to plant breeding
]
.author[
### Rafael Massahiro Yassue
]
.institute[
### GDM Seeds
]
.date[
### September 3, 2025
]

---

class: animate__animated, animate__fadeIn, has-logo

# Outline

1. Phenomics and Genetic Gain  
2. Case Studies  
3. Phenomics vs. Genomics  
4. Applications, Limitations, and Challenges
5. Phenomics Data Flow

---
class: animate__animated, animate__fadeIn
# Phenomics and Genetic Gain

$$
\Delta G = \frac{\color{red}{i} \cdot \color{red}{r_a} \cdot \color{red}{\sigma_a}}{\color{red}{L}}
$$

`\begin{aligned}
\Delta G &amp; : \text{genetic gain per selection cycle} \\
\color{red}{i} &amp; : \text{selection intensity} \\
\color{red}{r_a} &amp; : \text{selection accuracy (improved with high-throughput phenotyping)} \\
\color{red}{\sigma_a} &amp; : \text{additive genetic standard deviation} \\
\color{red}{L} &amp; : \text{length of the selection cycle (years)}
\end{aligned}`

--


&lt;table class="table table-striped table-hover table-condensed" style="font-size: 14px; width: auto !important; margin-left: auto; margin-right: auto;"&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:center;"&gt; Component &lt;/th&gt;
   &lt;th style="text-align:center;"&gt; How Phenomics Increases Genetic Gain &lt;/th&gt;
   &lt;th style="text-align:center;"&gt; Effect on Genetic Gain &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:center;width: 5em; font-weight: bold;"&gt; `\(i, r_{a}\)` &lt;/td&gt;
   &lt;td style="text-align:center;width: 20em; "&gt; Reduces experimental error &lt;/td&gt;
   &lt;td style="text-align:center;width: 20em; "&gt; ↑ Heritability → ↑ Accuracy &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:center;width: 5em; font-weight: bold;"&gt; `\(i, r_{a}\)` &lt;/td&gt;
   &lt;td style="text-align:center;width: 20em; "&gt; Enables repeated temporal measurements &lt;/td&gt;
   &lt;td style="text-align:center;width: 20em; "&gt; ↑ Precision of estimates &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:center;width: 5em; font-weight: bold;"&gt; `\(\sigma_{a}\)` &lt;/td&gt;
   &lt;td style="text-align:center;width: 20em; "&gt; Increases the number of plants measured &lt;/td&gt;
   &lt;td style="text-align:center;width: 20em; "&gt; ↑ Statistical robustness and ↑ genetic variance &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:center;width: 5em; font-weight: bold;"&gt; `\(L\)` &lt;/td&gt;
   &lt;td style="text-align:center;width: 20em; "&gt; Early selection &lt;/td&gt;
   &lt;td style="text-align:center;width: 20em; "&gt; Genetic correlation between traits (e.g., root angle ↔ resilience) or early evaluation (e.g., disease resistance) &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;



---
class: animate__animated, animate__fadeIn
name: phenomics-combined

# Phenomics: More Genetic Gain, Lower Cost

- **Use of secondary traits** correlated with the main trait

--

- **High initial investment** in sensors &amp; infrastructure

--

- **Low-cost platforms** enable large-scale data capture

--

- **Automation** reduces labor costs

--

- **No reagent costs**

--

- **&lt;span style="color:red; font-weight:bold;"&gt;Cost per data point drops drastically as scale increases!&lt;/span&gt;**

---
class: animate__animated, animate__fadeIn, inverse, center, middle

# Case Study: Integrating Genomics and High-Throughput Phenotyping in the Genotype × Plant Growth-Promoting Bacteria Interaction

---
class: animate__animated, animate__fadeIn

# Materials and Methods

+ Greenhouse conditions  
+ Early maize development stage (six fully expanded leaves)  
+ 360 inbred lines  
+ 13,826 SNPs  
+ Two managements, both under nitrogen limitation:  
  - **B+** Synthetic population of 4 PGPR (Plant Growth-Promoting Rhizobacteria):  
      * *Bacillus thuringiensis* RZ2MS9  
      * *Delftia* sp. RZ4MS18  
      * *Pantoea agglomerans* 33.1  
      * *Azospirillum brasilense* Ab-v5  
  - **B-** Without PGPR  
+ 2 replicates  
+ Total of 1,512 plots

---
class: animate__animated, animate__fadeIn

# Why Use High-Throughput Phenotyping?
&lt;img src="figs/HTP.png" width="85%" style="display: block; margin: auto;" /&gt;
  
---

class: animate__animated, animate__fadeIn

# A Low-Cost Greenhouse High-Throughput Phenotyping Platform

&lt;img src="figs/Screen Shot 2022-01-05 at 7.54.52 PM.png" width="75%" style="display: block; margin: auto;" /&gt;
&lt;p style="font-family: times, serif; font-size:11pt; font-style:italic"&gt;
Figure. Representation of the low-cost High-Throughput Phenotyping platform.
&lt;/p&gt;


---
class: animate__animated, animate__fadeIn

# Thermal and Multispectral Image Collection

&lt;img src="figs/RGB_colection.gif" width="50%" /&gt;&lt;img src="figs/thermal_coletion.gif" width="50%" /&gt;
&lt;p style="font-family: times, serif; font-size:11pt; font-style:italic"&gt;
Figure. Representation of RGB (left) and thermal (right) image acquisition using the low-cost High-Throughput Phenotyping (HTP) platform.
&lt;/p&gt;

---
class: animate__animated, animate__fadeIn

# Data Acquisition

&lt;img src="figs/HTP/fig01.png" width="90%" style="display: block; margin: auto;" /&gt;
&lt;p style="font-family: times, serif; font-size:11pt; font-style:italic"&gt;
Figure. Design of the greenhouse-based high-throughput phenotyping platform.
&lt;/p&gt;


---
class: animate__animated, animate__fadeIn

# Data processing 
&lt;img src="figs/fig02.png" width="75%" style="display: block; margin: auto;" /&gt;
&lt;p style="font-family: times, serif; font-size:10pt; font-style:italic"&gt;
Figure. Summary of multispectral image processing. (A) image acquisition; (B) orthomosaic assembly; (C) plot clipping; (D) canopy cover (CC); (E) dense point cloud; (F) plant height (PH) calculated from differences between the dense point cloud with plants and the dense point cloud with only the pot; (G) dense point cloud after applying a filter to remove the background; (H) dense point cloud for each plot; (I) 3D reconstruction of the dense point cloud to obtain canopy volume; and (J) f(biomass) obtained from the product of plant height (PH) and CC.
&lt;/p&gt;


---
class: animate__animated, animate__fadeIn
&lt;img src="figs/TPPJ.png" width="120%" style="display: block; margin: auto;" /&gt;

---
# Plant Growth
&lt;img src="figs/PG.png" width="60%" style="display: block; margin: auto;" /&gt;
&lt;p style="font-family: times, serif; font-size:11pt; font-style:italic"&gt;
Figure. Growth patterns of genotypes across maize development with (B+) or without (B–) plant growth-promoting bacteria inoculation. The blue and red dashed lines represent the means of B+ and B– managements, respectively. 
&lt;/p&gt;

---
class: animate__animated, animate__fadeIn

# Phenotypic and Genomic Correlations
&lt;img src="figs/table1.png" width="100%" style="display: block; margin: auto;" /&gt;
---
class: animate__animated, animate__fadeIn

# Shovelomics 

&lt;img src="figs/molbreed.png" width="100%" style="display: block; margin: auto;" /&gt;
---
class: animate__animated, animate__fadeIn
# Multi-phenotype Genomic prediction
&lt;img src="figs/MB/temp.png" width="70%" style="display: block; margin: auto;" /&gt;
&lt;p style="font-family: times, serif; font-size:11pt; font-style:italic"&gt;
Figure. Distribution of the significant SNPs from the GWAS analysis across the chromosomes for eleven traits and treatments (B+, B−, Delta).
&lt;/p&gt;



---
class: animate__animated, animate__fadeIn
# Multi-phenotype Genomic prediction
&lt;img src="figs/mol_1.png" width="70%" style="display: block; margin: auto;" /&gt;
&lt;p style="font-family: times, serif; font-size:11pt; font-style:italic"&gt;
Figure. Genomic prediction using GBLUP, GBLUP_GMS (GBLUP based on marker selection via GWAS), and GBLUP_MAS (GBLUP + marker-assisted selection) for various traits and managements.
&lt;/p&gt;


---
class: animate__animated, animate__fadeIn
# Hyperspectral Image Acquisition

&lt;img src="figs/fig01_da.png" width="90%" style="display: block; margin: auto;" /&gt;
&lt;p style="font-family: times, serif; font-size:11pt; font-style:italic"&gt;
Figure. Summary of data acquisition and processing. A) benchtop hyperspectral camera used for data collection; B) region of interest of the last fully expanded leaf; C) image cube; and D) image mask.
&lt;/p&gt;

---
class: animate__animated, animate__fadeIn
&lt;img src="figs/CS.png" width="100%" style="display: block; margin: auto;" /&gt;
---
class: animate__animated, animate__fadeIn
# Classification of Inoculation with Plant Growth-Promoting Bacteria
&lt;img src="figs/1-3.png" width="50%" style="display: block; margin: auto;" /&gt;
&lt;p style="font-family: times, serif; font-size:11pt; font-style:italic"&gt;
Figure. Classification accuracy of inoculation status (B+ and B−) using 150 hyperspectral bands.
&lt;/p&gt;

---
class: animate__animated, animate__fadeIn
# Genomic, Phenomic, and Multi-omics Prediction

&lt;img src="figs/1-4.png" width="80%" style="display: block; margin: auto;" /&gt;
&lt;p style="font-family: times, serif; font-size:11pt; font-style:italic"&gt;
Figure. Predictive ability of plant height (PH), stem diameter (SD), and shoot dry mass (SDM) using phenomic, genomic, and multi-omics prediction models under each management condition (B+ and B−).
&lt;/p&gt;

---
class: animate__animated, animate__fadeIn
&lt;img src="figs/PD.png" width="100%" style="display: block; margin: auto;" /&gt;
---
class: animate__animated, animate__fadeIn
# Phenome-wide association studies (PheWAS)
&lt;img src="figs/2-6.png" width="100%" style="display: block; margin: auto;" /&gt;
&lt;p style="font-family: times, serif; font-size:11pt; font-style:italic"&gt;
Figure. PheWAS for the 10 most frequent SNPs in models for plant height (PH), stem diameter (SD), and shoot dry mass (SDM) under B+ management.
&lt;/p&gt;
---
class: animate__animated, animate__fadeIn
# Multi-phenotype GWAS
&lt;img src="figs/2.gif" width="70%" style="display: block; margin: auto;" /&gt;
&lt;p style="font-family: times, serif; font-size:11pt; font-style:italic"&gt;
Figure. Multi-phenotype GWAS for the 281 hyperspectral phenotypes under B+ management.
&lt;/p&gt;

---
class: animate__animated, animate__fadeIn
# Thermal Camera in Disease Identification
&lt;img src="figs/758CFDF0-42F4-4FEB-ACC0-44973111C347.JPG" width="100%" style="display: block; margin: auto;" /&gt;

---
class: animate__animated, animate__fadeIn
# Thermal Camera in Disease Identification
&lt;img src="figs/plot_temp.png" width="120%" style="display: block; margin: auto;" /&gt;


---
class: animate__animated, animate__fadeIn, inverse, center, middle

# Applications, Factors Influencing Accuracy, and Limitations of the Methodology

---

class: animate__animated, animate__fadeIn
# Phenomics vs Genomics

&lt;div style="transform: scale(0.75); transform-origin: top center;"&gt;

&lt;table style="width:100%; border-collapse: collapse;"&gt;
  &lt;thead&gt;
    &lt;tr&gt;
      &lt;th style="border: 1px solid #ddd; padding: 6px; text-align: left;"&gt;Aspect&lt;/th&gt;
      &lt;th style="border: 1px solid #ddd; padding: 6px; text-align: left;"&gt;Genomic Selection (GS)&lt;/th&gt;
      &lt;th style="border: 1px solid #ddd; padding: 6px; text-align: left;"&gt;Phenomics / HTP&lt;/th&gt;
    &lt;/tr&gt;
  &lt;/thead&gt;
  &lt;tbody&gt;
    &lt;tr&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;Prediction basis&lt;/td&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;Molecular markers (SNPs)&lt;/td&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;High-resolution phenotypic traits&lt;/td&gt;
    &lt;/tr&gt;
    &lt;tr&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;Estimated effects&lt;/td&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;Genetic, mainly additive&lt;/td&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;Phenotypic variance; may include environmental effects and possible overestimation of accuracy in CV and validation&lt;/td&gt;
    &lt;/tr&gt;
    &lt;tr&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;When applied&lt;/td&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;Can be done &lt;b&gt;before field evaluation&lt;/b&gt;&lt;/td&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;Requires growing plants&lt;/td&gt;
    &lt;/tr&gt;
    &lt;tr&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;Accuracy&lt;/td&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;Depends on training population, years, panel size&lt;/td&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;High for &lt;b&gt;physiological and structural traits&lt;/b&gt;&lt;/td&gt;
    &lt;/tr&gt;
    &lt;tr&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;Cost per individual&lt;/td&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;&lt;b&gt;Medium&lt;/b&gt;: Laboratory + reagents&lt;/td&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;High initial cost, &lt;b&gt;drops with scale&lt;/b&gt;, no reagent cost&lt;/td&gt;
    &lt;/tr&gt;
    &lt;tr&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;Cycle speed&lt;/td&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;&lt;b&gt;Very fast&lt;/b&gt;, enables early selection&lt;/td&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;&lt;b&gt;Fast&lt;/b&gt;, but depends on phenological stage&lt;/td&gt;
    &lt;/tr&gt;
    &lt;tr&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;Traits&lt;/td&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;Prediction only for known traits&lt;/td&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;Inference of new physiological and structural traits&lt;/td&gt;
    &lt;/tr&gt;
    &lt;tr&gt;
      &lt;td style="border: 1px solid #ddd; padding: 6px;"&gt;Ideal integration&lt;/td&gt;
      &lt;td colspan="2" style="border: 1px solid #ddd; padding: 6px;"&gt;
        &lt;b&gt;GS + Phenomics → Multi-omics Integration:&lt;/b&gt; Increases accuracy, robustness, and genetic gain.
      &lt;/td&gt;
    &lt;/tr&gt;
  &lt;/tbody&gt;
&lt;/table&gt;

&lt;/div&gt;


---
class: animate__animated, animate__fadeIn

# Generation of Orthomosaics Using Limited-Resolution Images

&lt;img src="figs/problem/Screen Shot 2022-01-05 at 9.05.27 PM.png" width="100%" style="display: block; margin: auto;" /&gt;
&lt;p style="font-family: times, serif; font-size:11pt; font-style:italic"&gt;
Figure. Several attempts at orthomosaic assembly and image stitching using thermal images.
&lt;/p&gt;
---
class: animate__animated, animate__fadeIn

# HTP challenges
&lt;img src="figs/problem/1.png" width="100%" style="display: block; margin: auto;" /&gt;
&lt;p style="font-family: times, serif; font-size:11pt; font-style:italic"&gt;
Figure. Representation of different shapefiles.
&lt;/p&gt;

---
class: animate__animated, animate__fadeIn

# Generation of Shapefiles and Plot Delimitation
&lt;img src="figs/problem/2.png" width="199%" style="display: block; margin: auto;" /&gt;
&lt;p style="font-family: times, serif; font-size:11pt; font-style:italic"&gt;
Figure. Representation of different shapefiles.
&lt;/p&gt;
---
class: animate__animated, animate__fadeIn

# Generation of Shapefiles and Plot Delimitation
&lt;img src="figs/problem/3.png" width="100%" style="display: block; margin: auto;" /&gt;
&lt;p style="font-family: times, serif; font-size:11pt; font-style:italic"&gt;
Figure. Representation of different shapefiles.
&lt;/p&gt;

---
class: animate__animated, animate__fadeIn

# Generation of Shapefiles and Plot Delimitation
&lt;img src="figs/problem/4.png" width="100%" style="display: block; margin: auto;" /&gt;
&lt;p style="font-family: times, serif; font-size:11pt; font-style:italic"&gt;
Figure. Representation of different shapefiles.
&lt;/p&gt;

---
class: animate__animated, animate__fadeIn

# Generation of Shapefiles and Plot Delimitation
&lt;img src="figs/problem/4.png" width="100%" style="display: block; margin: auto;" /&gt;
&lt;p style="font-family: times, serif; font-size:11pt; font-style:italic"&gt;
Figure. Representation of different shapefiles.
&lt;/p&gt;

---
class: animate__animated, animate__fadeIn

# Factors Impacting the Accuracy of High-Throughput Phenotyping

&lt;div style="transform: scale(0.75); transform-origin: top center;"&gt;

| **Factor**                  | **Description**                                                                 |
| :-------------------------- | :------------------------------------------------------------------------------ |
| **Sensor Quality**          | Higher-resolution sensors capture more detailed data                            |
| **Environmental Conditions**| Climate, lighting, and temperature variations affect accuracy                    |
| **Phenotypic Diversity**    | Greater genetic variability in plants can generate more complex measurements     |
| **Data Processing**         | Quality of filtering, normalization, and data analysis algorithms                |
| **Calibration**             | Proper calibration of sensors ensures consistency and reduces errors             |
| **Sampling Density**        | More frequent sampling improves temporal resolution                              |
| **Data Fusion**             | Combining multiple sensor types increases accuracy in trait detection            |
| **Platform Stability**      | Fixed platforms produce more consistent and reliable data                        |
| **User Expertise**          | Knowledge of operation and interpretation influences reliability                 |


---
class: animate__animated, animate__fadeIn

# Current Challenges

&lt;div style="transform: scale(0.70); transform-origin: top left; overflow: auto;width: 150%; height: 100%"&gt;

| **Challenge Type**       | **Technical Challenges**                                        | **Biological Challenges**                                 |
| :----------------------- | :-------------------------------------------------------------- | :------------------------------------------------------- |
| **High Costs**           | Expensive equipment, especially advanced sensors                | Making sensors more accessible for broad use             |
| **Data Storage**         | Large data volumes require robust storage and processing        | Turning data into useful information for plant breeding  |
| **Calibration**          | Difficulties calibrating sensors due to environmental variables | Standardizing biological traits across different species |
| **Accuracy**             | Variation in sensor performance under different environments    | Biological variability between species may affect data consistency |
| **Data Integration**     | Difficulty integrating data from multiple sensor types           | Combining phenotypic and genetic data for better predictions |
| **Real-Time Analysis**   | Need to process large data volumes in real time                  | Complexity of analyzing dynamic biological processes in real time |
| **Platform Limitations** | Constraints in mobility and accessibility of platforms           | Adapting phenotyping methods to different crops and growth stages |
| **Scalability**          | Difficulty scaling HTP to large areas or multiple species        | Ensuring biological relevance at scale and across ecosystems |
| **Breeding Integration** | Defining breeding strategies when correlation with target trait is moderate |                                                          |
---
class: animate__animated, animate__fadeIn

# Phenomics Data Flow

&lt;img src="figs/workflow.png" width="100%" style="display: block; margin: auto;" /&gt;



---
class: animate__animated, animate__fadeIn

# In essence

- Phenomics increases genetic gain by improving accuracy and reducing costs  

- Affordable technologies allow large-scale use, even with limited budgets  

- Integration with genomics is the future: more robustness, more efficiency  

- Breeding strategies to maximize the accuracy of phenomics models  

- Multi-omics → the key to next-generation breeding

---
class: animate__animated, animate__fadeIn, middle

# Acknowledgments 

&lt;img src="figs/plot.jpg" width="100%" style="display: block; margin: auto;" /&gt;

---

class: animate__animated, animate__fadeIn

&lt;br /&gt; &lt;br /&gt; &lt;br /&gt; &lt;br /&gt; &lt;br /&gt; &lt;br /&gt;
#.center[Muito obrigado! ]

.right[[rafael.yassue@gmail.com](mailto:rafael.yassue@gmail.com?subject=%20Presentation)]
.right[https://rafaelyassue.github.io/cv-online/]
&lt;img src="figs/site2.png" width="50%" style="display: block; margin: auto 0 auto auto;" /&gt;

    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create();
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
